blueprints/flexible-topologies.md

```diff
@@ -0,0 +1,395 @@
+| cachegroup | string            | required    | the `short_name` of a cachegroup this node maps to in the Topology                                                                                                                                          |
+| parents    | array of integers | required    | zero-based indexes to other nodes in the Topology's `nodes` array, where the 1st element is for the *primary* parent relationship and the 2nd element is for the *secondary* parent relationship, and so on |
+
+API constraints:
+- a Topology must have at least 1 `node`; otherwise, it is useless
+- there cannot be multiple `nodes` for the same cachegroup in a Topology
+- `parents` must have 0, 1 or 2 elements, cannot contain duplicates, cannot contain the index of its own `node`, and cannot contain the index of `nodes` whose cachegroup is of type `EDGE_LOC`
+- leaf `nodes` must be cachegroups of type `EDGE_LOC`
+- all `nodes` in the Topology must be reachable -- i.e. a `node` is either a leaf (which would be an `EDGE_LOC`) or is a parent of at least one other node
+- a Topology cannot contain a cycle (through any combination of primary/secondary parent relationships)
+
+The following new endpoints will be required:
+
+##### `GET /topologies`
+
+response JSON:
+```JSON
+{ "response": [
+    {
+        "name": "foo",
+        "description": "a foo topology",
+        "nodes": [
+            {
+                "cachegroup": "child-cachegroup",
+                "parents": [1, 2]
+            },
+            {
+                "cachegroup": "parent-cachegroup",
+                "parents": []
+            },
+            {
+                "cachegroup": "secondary-parent-cachegroup",
+                "parents": []
+            }
+        ]
+    }
+]}
+```
+
+##### `POST /topologies`
+
+request JSON:
+```JSON
+{
+    "name": "foo",
+    "description": "a foo topology",
+    "nodes": [
+        {
+            "cachegroup": "child-cachegroup",
+            "parents": [1, 2]
+        },
+        {
+            "cachegroup": "parent-cachegroup",
+            "parents": []
+        },
+        {
+            "cachegroup": "secondary-parent-cachegroup",
+            "parents": []
+        }
+    ]
+}
+```
+
+response JSON:
+```JSON
+{
+    "alerts": [
+        {
+            "text": "topology was created successfully",
+            "level": "success"
+        }
+    ],
+    "response": {
+        "name": "foo",
+        "description": "a foo topology",
+        "nodes": [
+            {
+                "cachegroup": "child-cachegroup",
+                "parents": [1, 2]
+            },
+            {
+                "cachegroup": "parent-cachegroup",
+                "parents": []
+            },
+            {
+                "cachegroup": "secondary-parent-cachegroup",
+                "parents": []
+            }
+        ]
+    }
+}
+
+```
+
+##### `PUT /topologies?name=foo`
```

Comment:
are we using query params for PUT/DELETE? I'm fine with it, just thought we were not.

Reply:
A big push in API 2.0 was to cut down on duplicated endpoints which involved using query parameters instead of path parameters for a lot of things.

For my money, a path parameter would be fine, but then PUT with query parameters should not be supported, that way endpoints aren't duplicated. But more importantly, endpoints (like this one) should always support GET and `/foos/{{ID}}` shouldn't be handled exactly like `/foos?id={{ID}}`, which has historically never been the case. If no foo with ID=4 exists, then `GET /foos/4` should be a 404, while `/foos?id=4` should be an empty array. And actually, `/foos/{{ID}}` should never return an array while `/foos` should always return an array (for GET, at least).

Using query params for everything like this certainly cuts down on the amount of code you need to write, and especially since we currently have no real API standards I'm personally fine with that.

Reply:
Yes, that's all exactly right. The only thing is the guidelines recommend PATCH as well.